lib/ret_web/controllers/api/v1/media_controller.ex

```diff
@@ -10,29 +10,48 @@ defmodule RetWeb.Api.V1.MediaController do
     resolve_and_render(conn, url, 0)
   end
 
-  def create(conn, %{
-        "media" => %Plug.Upload{filename: filename, content_type: "application/octet-stream"} = upload,
-        "promotion_mode" => "with_token"
-      }) do
-    store_and_render_upload(conn, upload, MIME.from_path(filename), SecureRandom.hex())
+  def create(
+        conn,
+        %{"media" => %Plug.Upload{filename: filename, content_type: "application/octet-stream"} = upload} = params
+      ) do
+    convert_to_content_type = params |> Map.get("convert_to")
+    promotion_token = params |> promotion_token_for_params
+
+    store_and_render_upload(conn, upload, MIME.from_path(filename), convert_to_content_type, promotion_token)
   end
 
-  def create(conn, %{
-        "media" => %Plug.Upload{content_type: content_type} = upload,
-        "promotion_mode" => "with_token"
-      }) do
-    store_and_render_upload(conn, upload, content_type, SecureRandom.hex())
+  def create(conn, %{"media" => %Plug.Upload{content_type: content_type} = upload} = params) do
+    convert_to_content_type = params |> Map.get("convert_to")
+    promotion_token = params |> promotion_token_for_params
+
+    store_and_render_upload(conn, upload, content_type, convert_to_content_type, promotion_token)
   end
 
-  def create(conn, %{"media" => %Plug.Upload{filename: filename, content_type: "application/octet-stream"} = upload}) do
-    store_and_render_upload(conn, upload, MIME.from_path(filename))
+  defp promotion_token_for_params(%{"promotion_mode" => "with_token"}), do: SecureRandom.hex()
+  defp promotion_token_for_params(_params), do: nil
+
+  defp store_and_render_upload(conn, upload, content_type, convert_to_content_type, promotion_token)
+       when is_nil(convert_to_content_type) do
+    store_and_render_upload(conn, upload, content_type, promotion_token)
   end
 
-  def create(conn, %{"media" => %Plug.Upload{content_type: content_type} = upload}) do
-    store_and_render_upload(conn, upload, content_type)
+  defp store_and_render_upload(conn, upload, content_type, convert_to_content_type, promotion_token) do
+    case Ret.Speelycaptor.convert(upload, convert_to_content_type) do
+      {:ok, converted_path} ->
+        converted_upload = %Plug.Upload{
+          path: converted_path,
+          filename: upload.filename,
+          content_type: convert_to_content_type
+        }
+
+        store_and_render_upload(conn, converted_upload, convert_to_content_type, promotion_token)
+
+      _ ->
+        store_and_render_upload(conn, upload, content_type, promotion_token)
```

Comment:
Wouldn't it be better to error here, since we were asked to convert the upload?

Reply:
I don't think so, because (despite the probably poorly named API) the contract is "try to convert" -- for example, if you don't have the lambda configured, I want that to just gracefully degrade to not conversion. as a general facility this may not make sense, but for the specific thing we're using it for (the camera videos) it is def what we want.

Reply:
i'll rename the arg to `desired_format`